test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py

```diff
@@ -1,149 +1,178 @@
+
+    @classmethod
+    def default(cls):
+        return cls(0)
+
+    def __bool__(self):
+        return self > 0
+
+
+class uint(BasicValue, metaclass=BasicType):
+
     def __new__(cls, value, *args, **kwargs):
         if value < 0:
             raise ValueError("unsigned types must not be negative")
+        if cls.byte_len and value.bit_length() > (cls.byte_len << 3):
+            raise ValueError("value out of bounds for uint{}".format(cls.byte_len * 8))
         return super().__new__(cls, value)
 
+    def __add__(self, other):
+        return self.__class__(super().__add__(coerce_type_maybe(other, self.__class__, strict=True)))
+
+    def __sub__(self, other):
+        return self.__class__(super().__sub__(coerce_type_maybe(other, self.__class__, strict=True)))
+
+    @classmethod
+    def default(cls):
+        return cls(0)
+
 
 class uint8(uint):
     byte_len = 1
 
-    def __new__(cls, value, *args, **kwargs):
-        if value.bit_length() > 8:
-            raise ValueError("value out of bounds for uint8")
-        return super().__new__(cls, value)
-
 
 # Alias for uint8
-byte = NewType('byte', uint8)
+class byte(uint8):
+    pass
 
 
 class uint16(uint):
     byte_len = 2
 
-    def __new__(cls, value, *args, **kwargs):
-        if value.bit_length() > 16:
-            raise ValueError("value out of bounds for uint16")
-        return super().__new__(cls, value)
-
 
 class uint32(uint):
     byte_len = 4
 
-    def __new__(cls, value, *args, **kwargs):
-        if value.bit_length() > 32:
-            raise ValueError("value out of bounds for uint16")
-        return super().__new__(cls, value)
-
 
 class uint64(uint):
     byte_len = 8
 
-    def __new__(cls, value, *args, **kwargs):
-        if value.bit_length() > 64:
-            raise ValueError("value out of bounds for uint64")
-        return super().__new__(cls, value)
-
 
 class uint128(uint):
     byte_len = 16
 
-    def __new__(cls, value, *args, **kwargs):
-        if value.bit_length() > 128:
-            raise ValueError("value out of bounds for uint128")
-        return super().__new__(cls, value)
-
 
 class uint256(uint):
     byte_len = 32
 
-    def __new__(cls, value, *args, **kwargs):
-        if value.bit_length() > 256:
-            raise ValueError("value out of bounds for uint256")
-        return super().__new__(cls, value)
-
-
-def is_uint_type(typ):
-    # All integers are uint in the scope of the spec here.
-    # Since we default to uint64. Bounds can be checked elsewhere.
-    # However, some are wrapped in a NewType
-    if hasattr(typ, '__supertype__'):
-        # get the type that the NewType is wrapping
-        typ = typ.__supertype__
 
-    return isinstance(typ, type) and issubclass(typ, int) and not issubclass(typ, bool)
+def coerce_type_maybe(v, typ: SSZType, strict: bool = False):
+    v_typ = type(v)
+    # shortcut if it's already the type we are looking for
+    if v_typ == typ:
+        return v
```

Comment:
Should this return a new instance of the value rather than the same to avoid mutability issues?

Reply:
No, if we set a value/member, we want to get the proper python object into that spot. Only if it's not the same type, then we coerce it. But for lists etc, this only really means that the elements get a new container. And for basic types (uint, bool), it doesn't matter if we copy or not.